### ISSUE TITLE
minetest.after considered harmful

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#5488]
Based on my long experience with the Lua API, I think the `minetest.after` function is one of the most frustrating functions in the API.

There are many things wrong about this function:

* If you shutdown the server, all `minetest.after` timers are gone for good, making them useless for serious gameplay concepts
* `minetest.after` timers don't seem to work correctly when the game is paused in singleplayer mode
* You can't stop, pause or abort a `minetest.after` timer. Once it runs, there is no turning back. This makes timer management very painful

In practice, this function is not really useful and an endless source of bugs in mods and subgames because the timing is so confusing and inconsistent. In order to use this function for most real-world use cases, you have to add boilerplate checking code or even dump data into files, both are unacceptable solutions.

I think this function should eventually be phased out.

What Minetest needs is a proper timer API. The closest we have for “real” timers are node timers, but they only work on nodes.

I suggest to add a timer API with the following featuers:

* Basically all the features of node timers
* And a similar API like for node timers
* But it can be used independently of node timers
* Respects singleplayer pause
* You get a handle after creating a timer (like for node timers)  which you can use to further track or manipulate the timer
* Is NOT forgotten after shutdown

**Comment by bigfoot547** [comment]
Yeeeessss

**Comment by sofar** [comment]
well if it's lame then I'm not going to fix it.

**Comment by sofar** [comment]
@Wuzzy2 I can surely imagine something that meets your requirements, but, I am having a hard time understanding whether you're just criticizing the code because it forces you to do some homework, and you're just lazy (obviously I know you're not, but indulge me for a second).

> this function is not really useful

```
$/minetest-mods/ $ grep -r minetest.after ./*/*.lua|wc -l
90
```

I dunno, that seems pretty useful to me. Many mods are using it?

> If you shutdown the server, all minetest.after timers are gone for good, making them useless for serious gameplay concepts

Which is why you schedule your timers at init time? Or when the first player connects?

They're used for many, very serious tasks. So this is just a superlative argument. Obviously many mods have been able to work with the current API just fine.

>You can't stop, pause or abort a minetest.after timer. Once it runs, there is no turning back. This makes timer management very painful
>You also can't query the remaining time of an active timer

No, but you can easily set a flag and process it in the timer routine to prevent rescheduling. Doesn't seem to be a deal-breaking issue therefore.

> In practice, this function is not really useful and an endless source of bugs in mods and subgames because the timing is so confusing and inconsistent (I speak from experience). In order to use this function for most real-world use cases, you have to add boilerplate checking code or even dump data into files, both are unacceptable solutions.

> "Unacceptable"

> "endless source of bugs"

superlatives, again.

Sorry, it just sounds like bitching.

**Comment by sofar** [comment]
> I suggest to add a timer API with the following featuers:

> Basically all the features of node timers
> And a similar API like for node timers
> But it can be used independently of node timers

> All timers must be paused while in singleplayer pause

So all timers expire all at the same time when the game unpauses after being paused for half an hour? That sounds ... dangerous ...

> You get a handle after creating a timer (like for node timers) which you can use to further track or manipulate the timer

How does the handle work after a restart?

> Is NOT forgotten after shutdown

I can actually foresee such a timer system. But, funny thing, it would probably end up using `minetest.after()` under the hood to provide good performance :)

Persistency means we need to use some internal `modstorage` for timers, which is doable alright, but if a mod wants to keep a handle for a timer it would also have to maintain it's own reference to the handle in it's own `modstorage`. Not very clean if you ask me, but would be doable I think.

**Comment by HybridDog** [comment]
> So all timers expire all at the same time when the game unpauses after being paused for half an hour? 

l think so because [minetest.after uses minetest.get_us_time and not minetest.get_gametime.](https://github.com/minetest/minetest/blob/master/builtin/common/after.lua#L6)

**Comment by raymoo** [comment]
Something like this should allow metadata to be associated with a particular instance of a timer. Nodes can live without this since they have node metadata in addition to the timer, but a detached timer would need some other way to keep track.

**Comment by Wuzzy2** [comment]
>> this function is not really useful
>
> $/minetest-mods/ $ grep -r minetest.after ./*/*.lua|wc -l
> 90
>
> I dunno, that seems pretty useful to me. Many mods are using it?

What I mean with this is that it is not easy to use, there are many pitfalls and gotchas because of missing features. Key features such as persistance are missing.

>> If you shutdown the server, all minetest.after timers are gone for good, making them useless for serious gameplay concepts

> Which is why you schedule your timers at init time? Or when the first player connects?

Yes, but this only works when you work with custom files, how else would you be able to do this? I'm afraid this is really the only way to avoid “forgetting” the timers. Minetest does not give me any tools whatsoever to restore the timers of the previous session, I'm completely on my own.
Requiring mods to create non-standard files for very basic timer stuff is not good. You also don't need extra files for node timers. :-(

And even if you do this, you still have trouble. Since we cannot query the remaining time of a timer directly, we can only do this indirectly via OS timers. This is ugly and we are losing accuracy for no reason whatsoever.

> They're used for many, very serious tasks. So this is just a superlative argument. Obviously many mods have been able to work with the current API just fine.

They are just using `minetest.after` because it's basically the only way to do timer stuff *at all*. Of course you can still *kinda* work with `minetest.after`, but this doesn't explain away all the problems.

> I can actually foresee such a timer system. But, funny thing, it would probably end up using minetest.after() under the hood to provide good performance :)

Well, as long the missing features are implemented, I'm probably happy with any solution.

>> You can't stop, pause or abort a minetest.after timer. Once it runs, there is no turning back. This makes timer management very painful
>> You also can't query the remaining time of an active timer
> No, but you can easily set a flag and process it in the timer routine to prevent rescheduling. Doesn't seem to be a deal-breaking issue therefore.

And how do you query the remaining time of a timer? And how do you know a timer still exists? IMO this is pretty basic stuff which I can't do.

> So all timers expire all at the same time when the game unpauses after being paused for half an hour? That sounds ... dangerous ...

No! Why would the timers do such a crazy thing? The timers are PAUSED. Meaning, they do not count down. When the game is unpaused, the timers just continue to run normally.
Example: If a timer with 4 seconds left is paused because of singleplayer pause, it will of course keep this time no matter how long the singleplayer pause lasts. When the game unpauses, so do the timers, so the 4 second timer runs with its 4 seconds left.

>> "Unacceptable"
>> "endless source of bugs"
>
> superlatives, again.
> Sorry, it just sounds like bitching.

I just speak from experience with my own mods. Examples:

* Persistance is unneccessary hard to do and error-prone, because you have to juggle with custom files. This introduces avoidable complexity into the code. And I think persistance of timers is a very basic thing. From experience, I pretty much always need persistance, because a server shutdown is a completely unpredictable event. Therefore, the fact that there is no-built in persistence makes it pretty painful to work with timers
* Because you can't query the remaining time of timers, you can only make guesses based on OS time
* `minetest.after` doesn't seem to be paused in singleplayer pause, introducing a lot of unexpected behaviour into mods and it is next to impossible to mitigate
* I dislike the notion of moving too much core responsibility to mods. It's not because I'm lazy, it's because I think features which mods need over and over again must go into the engine. Asking modders to implement core features leads to redundancy

> Persistency means we need to use some internal modstorage for timers, which is doable alright, but if a mod wants to keep a handle for a timer it would also have to maintain it's own reference to the handle in it's own modstorage. Not very clean if you ask me, but would be doable I think.

Sounds semi-reasonable. But I have thought of a system where Minetest simply automatically saves all active timers into the world file, so the mod does not have to do anything. Similar to what Minetest does with player data: This data is also automatically saved, and the mods don't have to do anything. I think modstorage code would be overkill here.

As for the question how mods are supposed to get time
Maybe just add a simple function `minetest.get_timers()` to grab a list of all currently active timers? And maybe a callback for when Minetest has loaded them after a restart. That way, populating the list of timers in a mod will become trivial. :-)

**Comment by sofar** [comment]
@Wuzzy2 I'm reading all of it but I still only see "complaints" and not a single convincing reason to implement a more complex timer class.

Please, show me where you had to "kludge around" to make `minetest.after()` to work in your code, or better, show me code that is buggy because of `minetest.after()`.

I can't solve the problem if I don't understand what actually is the problem. Arguments like "it's too hard" is not a valid argument, since "too hard" may be different for me and you (you need to elaborate why it is too hard, preferably with code examples).

> Maybe just add a simple function minetest.get_timers() to grab a list of all currently active timers?

Uhh, I don't think we should let mods know about *all* timers, that sounds very dangerous. How does it know what timers belong to its own mod code and not another mod?

I think persistency is largely not a good thing anyway. What if somoene removes a mod, and now the timer calls the nonexistant lua code? We can't store the code either, since the mod may have been upgraded.

**Comment by sofar** [comment]
@HybridDog if the time resolution of get_game_time() is detailed enough we should just switch to that.

**Comment by sofar** [comment]
@HybridDog aaand no, it returns only `seconds` precision, so, it's useless for `minetest.after()`

**Comment by sofar** [comment]
#5500 makes singleplayer pause also pause `minetest.after()` scheduled timers.

**Comment by nerzhul** [comment]
#5500 is now merged

**Comment by Wuzzy2** [comment]
> #5500 makes singleplayer pause also pause minetest.after() scheduled timers.

Thank you! I tick one item off the TODO list. :-)

> I think persistency is largely not a good thing anyway.

(If you meant this sentence serious, read on. If not, you can ignore the following wall of text. xD)

I think timer persistency is a very important thing. It's strange we have it for node timers but not for `minetest.after`. Your opinion does not make sense: If persistency is a bad thing, why do we have it for node timers in the first place?

Examples where persistent timers are needed:

* Temporary status effects, like in the `playereffects` mod. Examples:
    * Health regeneration (e.g. +2 HP every 2 seconds)
    * Run faster for 40 seconds
    * Jump higher for 10 seconds
    * Become immortal for 5 seconds
    * …
* Barking and regeneration after eating PB&J Pup (if we assume for a moment that we do NOT want to lose this regeneration after a restart)
* You use a rare item which has an effect (e.g. teleportation) 10 seconds after using it
* In short, ANY gameplay-relevant event which requires a timer (and thus must use `minetest.after`)
* The odd one: You punch a bomb block and it explodes after 5 seconds. (Surprise! This is already properly persisted thanks to node timers. :-))

If you just use `minetest.after` for any of the above, you will lose an important part of the game state after a simple server restart (except for the bomb block). This is BAD because a server restart is a completely unpredidictable event which can potentionally seriously break the gameplay (depending on the circumstances and mods).
For example, if you used the rare teleporter item, then the server suddenly restarts, you just wasted an item and won't get teleported.
For most other gameplay-relevant things (nodes, health, carried items), the game state is preserved automatically. This is not the case for pending timers, which are thrown away. IMO it does not make sense to throw this information away. It is reasonable for modders to expect that pending timers are persisted as well.

Can we at least agree that the pending timers are actually part of the game state and thus MUST be persisted?

If timers would be persisted, you can now safely use the teleporter item: Player uses it, the 10s timer starts, after 5s the server suddenly shuts down, the remaining 5s are *somehow* stored in the world files. Then the server boots up again, the 5s timer is restored and after 5s you actually teleport. For the game, it's as if the server has never been down. In one sense, the game could be considered to be “frozen” while it was shutdown.

> Please, show me where you had to "kludge around" to make minetest.after() to work in your code, or better, show me code that is buggy because of minetest.after().

OK. In `playereffects`, I mostly worked around the fact that timers are not persisted by Minetest by just manually putting all timer information into a file in the world directory and loading it afterwards. If I don't do this, players would automatically lose all status effects after a restart. The code “works”, but I dislike it and it even has a bug.

This is the relevant code:

* This “do” routine for loading, is called very early after start:
http://repo.or.cz/minetest_playereffects.git/blob/0f02d5e54b54b5738f2a9927681377423a5691fa:/init.lua#l48

* `playereffects.save_to_file()` for saving, is called on shutdown:
http://repo.or.cz/minetest_playereffects.git/blob/0f02d5e54b54b5738f2a9927681377423a5691fa:/init.lua#l310

I don't like this code at all because **it feels a lot like boilerplate** to me, because there are many mods which like to have persistent timers. If you recall my examples from above, you will notice that not just `playereffects` would have to add this boilerplate code to be able to actually have persistent timers, but ANY mod which wants to use a timer which is supposed to trigger ANY gameplay-related thing. **If all such mods would use this boilerplate, we get a lot of redundant code.**

In `playereffects.save_to_file()`, I also absolutely hate these bits of code:

    local save_time = os.time()
    
(…)

    new_repeat_duration = effect.repeat_interval_time_left - os.difftime(save_time, effect.repeat_interval_start_time)

Explanation: To save the timers, I have to know how much time is remaining for each of them. Sadly, Minetest does not have any API for this. So this code is basically a bad ugly workaround around this by using `os.time` to remember the start time and the time of shutdown, then using `os.difftime` to get the difference.

This doesn't feel right. First, it's using `os` instead of `minetest`. `os.time` only has a precision in the seconds. **It's guessing and “praying” that the calculated time actually matches the REAL remaining time**, so this introduces some inaccuracy for no reason.
I may have used a more precise function but I still wouldn't trust it so much. Precision is not the only problem. What's worse, **this hack falls on its face when the player used the singleplayer pause at any point while the status effect was still active. Then, the timer will be completely wrong.** :-( And I don't see any way with the current API to work around this. The only solution seems to be to fix the API.

And I would be very glad as soon I can nuke this terrible hack. :-)

> What if somoene removes a mod, and now the timer calls the nonexistant lua code? We can't store the code either, since the mod may have been upgraded.

> Uhh, I don't think we should let mods know about all timers, that sounds very dangerous. How does it know what timers belong to its own mod code and not another mod?

I think removing a mod is always going into the danger zone, at least for some mods, and we should not expect this to be a common event.
For such a timer, I suggest to just throw it away. If Minetest finds a timer but doesn't match any code anymore, throw it away. Yes, this will lose some of the game state but this is expected when removing a mod. Keeping this timer would probably also lead to much confusion.

Welllll, I admit my suggestions for HOW to actually persist the timers weren't so great, but I hope it is clear now that it is a GOOD thing to consider timers to be part of the game state. Maybe we can find another (a better) way to actually save the timers (I think it's possible, it works for node timers, so why not for “global” timers?).

But now I have written so much text and I'm a bit tired and guess I will think and talk about this later. ;-)

**Comment by rubenwardy** [comment]
Persisting timers shouldn't been done as minetest.after, but as another API that uses minetest.after. The problem with minetest.after is that you pass a function directly, so that would have to be serialised.

I suggest thing like:

```lua
minetest.register_on_event("pbjpup:buf_tick", function(put, params, here)
    -- do stuff
end)

minetest.event_after(0.5, "pbjpup:buf_tick", put, params, here)

minetest.event_interval(0.5, "pbjpup:buf_tick", put, params, here)
```

**Comment by beyondlimits** [comment]
I haven't read the whole thread but as a hint I'll give you sample how JavaScript in your browser does this:

```javascript
var t = setTimeout(function() {...}, 1000); // callback is called once, time given in milliseconds
var i = setInterval(function() {...}, 1000); // callback is called periodically, time given in milliseconds
...
clearInterval(i);
clearTimeout(t);
```

**Comment by raymoo** [comment]
@beyondlimits That doesn't look like it does persistence though? Maybe you didn't read the whole thread, but even the OP specifies persistence.

**Comment by beyondlimits** [comment]
Yeah, what I posted didn't mean to support persistence out of the box, Only to point out that:

- millisecond (or something better than second) timer precision might be good
- cancellable timers might be good.

**Comment by raymoo** [comment]
@beyondlimits Alright, that makes sense.

**Comment by numberZero** [comment]
Nice idea. And @rubenwardy explained how the API might look like. Although such detached timers may be a bit problematic, as the target object, be it a player, an entity, or whatever, may disappear on restart. So that’s mostly suitable for truly global events.

By the way, that *can* be done in a mod, using `register_globalstep` and mod storage.

**Comment by raymoo** [comment]
For object timers, you could stop them unless the object is around, or do something like node timers where you trigger it once the object loads and pass the actual time passed.

Also I don't think there's any argument over whether it could already be done in a mod, only about whether the functionality should be included for convenience and less feature duplication. For example, `minetest.after` itself could be and is implemented using a globalstep callback, but it's still provided by the engine as a convenience.

**Comment by paramat** [comment]
> Persisting timers shouldn't been done as minetest.after, but as another API that uses minetest.after

Yes, .after seems mainly to be for short delays.
Another possibility is a limited duration timer that delays shutdown if still running (like a lua mapgen does), to avoid the need for persistence.

**Comment by numberZero** [comment]
> limited duration timer that delays shutdown if still running

then its construction may fail. otherwise, it would be possible to repeatedly create such timers to prevent (clean) shutdown for unlimited time.

**Comment by numberZero** [comment]
@raymoo using features I noticed, it may be implemented in a clean and sane way, without all that horrible temporary files, os.time() calls, etc. So I think this should be implemented as a library (mod adding no content) first, while it might be included in the core later.

**Comment by sofar** [comment]
@Wuzzy2 

https://github.com/minetest-mods/timer

Please review this mod and see if it addresses your concerns. If this is what solves your problems, we can consider merging it into `builtin` as an additionally available class for mods to use.

**Comment by Wuzzy2** [comment]
Looks interesting. I am currently not in my Minetest phase, so I plan full review and an actual test for May 2017. Feature-wise, it does not look it is missing anything. But this is not my final word. I especially have to test that these timers actually work out of the box with restarts and pauses.

I am not sure why this is supposed to be done in `builtin` and not directly in the engine, but since it could become part of the official API either way, I won't hold string opinions about this.
Do you intend this to be a complete replacement to `minetest.after` or is it just supplemental?
Looking superficially at it, it looks it can do everything which `minetest.after` does, but in a much saner way. So I think if this timer mod turns out to be good, we might even remove or at least deprecate `minetest.after`.

Will the fact that this is Lua instead of C++ have a *serious* impact on precision?
I see you use `globalstep` for timers. As far I know, a `globalstep` can occour very chaotically in extreme situations. I wonder if this have an impact on the usability/performance/reliability of timers. Need to test.

**Comment by sofar** [comment]
> Do you intend this to be a complete replacement to minetest.after or is it just supplemental?
> Looking superficially at it, it looks it can do everything which minetest.after does, but in a much saner way. So I think if this timer mod turns out to be good, we might even remove or at least deprecate minetest.after.

I don't think it will deprecate minetest.after, since the benefits of not saving any state makes `after` very light weight and therefore cheap.

> Will the fact that this is Lua instead of C++ have a serious impact on precision?

unlikely, because even if ran from C++ you'd probably only process time events every server step.

**Comment by beyondlimits** [comment]
> I am not sure why this is supposed to be done in builtin and not directly in the engine

From my experience if a bug is found in the script, in most cases it's easier to fix (you just modify the code and test it immediately) that if it would be C++ code (especially if you work on Windows).

**Comment by raymoo** [comment]
Currently `minetest.after` is not lightweight in some cases due to a performance bug

**Comment by Wuzzy2** [comment]
I have now finally looked at and tested @sofar's `timer` mod and it looks good to me. This is exactly the API I (and many other modders) need!

Now I can finally have timers that survive restarts WITHOUT having to jump through hoops (i.e. deal with it manually). It also accounts for singleplayer pauses (also VERY important!).

The fact that I have to keep a mod storage reference is not a problem.
It's also possible to create a non-persistent timer if you really need that. So there is nothing `minetest.after` can do that `timer` can't do as well. I think `timer` totally crushes `minetest.after`.

I urge core devs to give `timer` a try.

Also, `timer` easily ticks off all my checkboxes, thank you, sofar!

Oh, by the way: It seems now (in 0.4.17.1) that `minetest.after` currently accounts for singleplayer pauses as well, so at least that's no longer an issue with `minetest.after`, as written in the initial post. But `timer` is still an important tool because of persistence.

What will be a problem that after merging `timer`, there will be now two redundant ways to deal with timers. I believe `minetest.after` (sadly) has to stay for legacy support, there is no other way. :/ Maybe mark `minetest.after` as deprecated (at least for modding use) because with `timer` you can have non-persistent timers, too.

@paramat: I have a feeling that sofar is supportive.

PS: Sorry for having waited far too long for the review! :( I really have no excuse here, it's totally my fault and I screwed up.

**Comment by numberZero** [comment]
> So I think this should be implemented as a library (mod adding no content) first, while it might be included in the core later.

As `mod` stands for `module`, that’s acceptable to have non-essential features in mods and not in the core. The advantage *should be* easier maintenance (the `timer` mod is not bug-free, see https://github.com/minetest-mods/timer/pull/1 =)

**Comment by Wuzzy2** [comment]
`timer` is not supposed to go into the engine (C++), only `builtin`.

Thanks for finding the bug in `timer`!

> Actually, would dependency tracking be automatic, the difference between core features and such mods would be hardly noticeable for the end user IMO, be he a player or a server operator.

Not true, the user needs to hunt down such mods. Also, if many games start to add “timer mods”, modders have to adapt to different API, which suchs.
It's better to do it right the first time, and with ONE working implementation in the core / `builtin`, than to do it with 3rd party mods which might be competing with each other.

Calling “mod” a “module” is just your interpretation, it seems. Is that meaning official? Because in official documentation (`lua_api.txt`) it is not clear/official that “mod” is really *meant* to be the short version for “module”.

I also like to stress that we're talking about *persistent timers* here, which I think are quite essential.

**Comment by paramat** [comment]
> What will be a problem that after merging timer, there will be now two redundant ways to deal with timers.

I disagree, because Lua code is slower and causes server lag when in operation. Any new timer feature should really be in the engine. I'm not even sure it's worth adding it to builtin, might as well go straight to an engine implementation and in the meantime just use the 'timer' mod.
A non-persistent timer is likely to be less intensive so is still very useful, most timer usage does not require persistence.
So, 'after' will probably not be redundant or qualify for deprecation even with a new engine timer feature.
> I have a feeling that sofar is supportive.

Yes, potentially, so removed the label.

**Comment by rubenwardy** [comment]
`minetest.after` is already in Lua. Saying that Lua is slow is just wrong and uniformed

**Comment by paramat** [comment]
Oops, it is i Lua, sorry (embarassed).

**Comment by sofar** [comment]
@Wuzzy2 

> Am I overlooking something here?

I assumed that people would register a function for each timer class, and so I haven't given this usage any thought yet. It could make sense, and, I'll need to consider it for a bit first (it's been a ... little bit since I wrote this code :))

**Comment by sofar** [comment]
@Wuzzy2 

The `Timer` class can use statically coded timers, or you can dynamically create timers and use the `key` value in the `data` struct to distinguish them. Make sure `key` remains unique for your mod. `fn` Is not an identifier, you can have many timers use the same `fn` value.

Either timers are transient (no `key` and `storage` defined) or they are identified by the `key` value.

For instance, if you have a timer events `foo` and `bar` you could create `timer_foo` and `timer_bar` keys in your storage using the `data` struct. Of course, you would also have to maintain a separate table with `timer_foo` and `timer_bar` somewhere so that after reloading from a game shutdown, you could re register each timer. 

Something like this:

```
local my_timers = {}
-- S == modstorage obj
for k, _ in S:stored_timers do
      my_timers = Timer(my_timer_fn, {
            storage = S,
            key = k,
            interval = 3.0,
            repeats = true})
end
```

**Comment by Wuzzy2** [comment]
> Of course, you would also have to maintain a separate table with timer_foo and timer_bar somewhere so that after reloading from a game shutdown, you could re register each timer.

So, I still have to do it “by hand”? Lol, but that would kind of miss the point of having an API in the first place. o_O

I also don't see a “data” struct in your example. Your example is confusing.

Is it really so complicated to attach custom data to a timer? In `minetest.after` it's actually much simpler.

**Comment by numberZero** [comment]
@Wuzzy2 That’s why the API shouldn’t be in the core: mod can be changed or replaced at any time, but the core is effectively frozen in that aspect.

**Comment by Wuzzy2** [comment]
...

That's why I want the timer API to be feature-complete and stable *before* getting merged. If the timer API has fewer features than `minetest.after`, it's obviously not an improvement and should not be merged *yet*.

But when the timer API has proven itself worthy, I see no reason for not pushing it to builtin.

**Comment by sofar** [comment]
@wuzzy2 

https://github.com/minetest-mods/timer#creating-a-timer

> So, I still have to do it “by hand”? Lol, but that would kind of miss the point of having an API in the first place. o_O
>
> I also don't see a “data” struct in your example. Your example is confusing.
> 
> Is it really so complicated to attach custom data to a timer? In minetest.after it's actually much simpler.

I meant `timerdata`. Please review the API documentation I linked above.

It seems you do not understand the design choices made in the Timer class and why it functions as it does and not "the way you want it to function". So, here is my utter sucky attempt to explain it to you: (if I fail it isn't my fault, at least I gave it a good try).

The Timer class is designed to accommodate arbitrary timers from mods in a safe way. This means that the timers should not be reloaded after a game reboot if the mod that created them are no longer loaded, it also means that if the mod has been upgraded and now the timer function is a different function, things can keep working.

For this very simple reason, the timer can not do the following:
1) store a function across server restart - the function must be able to be changed by the mod
2) store mod private data across restart outside of the mod - the mod must be able to change, delete or add private data to this timer across a restart

These 2 restrictions are critical, because without these the mods wouldn't be able to modify or upgrade timers across server restarts to handle new functionality or different code handlers.

From all this results that the timer is defined by the mod with a `fn` and a `timerdata` that has to refer back to the `mod_storage` of each mod to keep the timer data private to the mod. Then this leads to the design that each mod is responsible for identifying timers that it needs to initialize at module load time. This isn't an arbitrary restriction - it follows logically from the given restrictions listed above, as you can see.

Now, it is certainly possible that the `Timer` mod could maintain a list of timers for each mod, but it would still have to obtain a handle to the `mod_storage` object for each mod, therefore, it can certainly by hypothesized that there would be a convenience function that would retrieve and revive all stored timers, however, at the time of writing the timer mod I decided against it because IMHO it doesn't add any significant value and any mod that needs to perform upgrade or modification to the timerdata (use a different key etc) can't use it and it would make the function name permanently locked forever, which is IMHO the wrong approach. For that reason, I doubt that the the timer API will ever change much.

So, LT:DR: the timer class design has undergone significant serious thought, and your functionality ideas lead IMHO to an inferior design that is less robust and prone to upgrade limitations, which I intended to avoid.

**Comment by sofar** [comment]
@Wuzzy2 

> Is it really so complicated to attach custom data to a timer? In minetest.after it's actually much simpler.

That's because `minetest.after` doesn't even survive a restart, so you can attach transient objects to it without penalty. 

From a must-work-across-restart perspective, this is unwanted.

Hypothetical example: A mod that creates an "aoe" damage effect around the player. The player retain the effect through login/logout and server restart.

Results:
- If this timer runs when the player logs out, it must restart when the player logs back in
- therefore, if the player logs in after a reboot, you can't refer to the `player` object. You must refer back tot he player `name` in the timer somehow.
- the mod therefore must keep a list of "aoe" "timers" for "players" and when a player logs in, the mod itself has to create the timer for the player at that time.

Note that restoring the timer doesn't happen when the server restarts. Mods can control this, but the `Timer` class itself doesn't know that the timer is attached to an entity, a node, a player or something just ephemeral. All of this requires the mod to extensively maintain the timer state already. And the timer class provides the right building blocks to support all of these use cases.

Adding more private data to `timerdata` wouldn't make any significant change. Through the `key` member the mod can already do everything it wants, at the cost of making the mod do a tiny bit more work to retrieve additional data. We could add an extra private member to the `timerdata` struct passed to the timer mod during creation time, of course, and it could indeed be modified and changed, and that may be an interesting extension, but it is not a significant loss of functionality.

**Comment by paramat** [comment]
Any core dev support for an additional timer API? (i'm neutral).

**Comment by raymoo** [comment]
It seems to me that if a mod can upgrade all of its timer data at mod startup, it can also do so on the fly when it sees outdated metadata.

**Comment by sofar** [comment]
> It seems to me that if a mod can upgrade all of its timer data at mod startup, it can also do so on the fly when it sees outdated metadata.

Yes but keeping as much private data out of the timer is good API design. You should only have 1 handle and not 2, or 3. The handle allows the mod at startup to create the proper metadata and upgrade any private data itself stores. Moving this metadata to a different mod through a secondary API just makes it fragile and complex.

**Comment by sofar** [comment]
@paramat neutral myself. The code itself can easily be kept in a mod and imported by games that require it. I can go either way.
